src/Application/Kernel.php

```diff
@@ -161,4 +149,57 @@ private function initDiscovery(Container $container): void
             $discovery->storeCache();
         }
     }
+
+    private function discoverInstalledPackageLocations(): void
+    {
+        $composerPath = path($this->root, 'vendor/composer');
+        $installed = path($composerPath, 'installed.json');
+
+        if (is_file($installed)) {
+            $installed = $this->loadJsonFile($installed);
+
+            foreach ($installed['packages'] ?? [] as $package) {
+                $packagePath = realpath(path($composerPath, $package['install-path']));
+
+                if (isset($package['require']['tempest/framework'], $package['autoload']['psr-4'])
+                    && is_array($package['autoload']['psr-4'])
+                ) {
+                    foreach ($package['autoload']['psr-4'] as $namespace => $namespacePath) {
+                        $namespacePath = path($packagePath, $namespacePath);
+                        $this->addDiscoveryLocation($namespace, $namespacePath);
+                    }
+                }
+            }
+        }
+    }
+
+    private function discoverTempestNamespaces(): void
+    {
+        $composer = $this->loadJsonFile(path($this->root, 'composer.json'));
+
+        $autoloads = [$composer['autoload']['psr-4'] ?? []];
+        if (env('ENVIRONMENT') === 'production') {
+            $autoloads[] = $composer['autoload']['psr-4'] ?? [];
+        }
+        $autoload = array_merge(...$autoloads);
```

Comment:
What's happening here exactly? Looks a bit weird to me that we first store everything in `$autoloads` and then merging it? Are you trying to flatten the array?

Reply:
The original functionality that came from using `autoload_psr4.php` was _all_ the defined namespaces are added to discovery - including those in `"autoload-dev"`.

i wanted to preserve that original logic, but it doesn't make sense (to me) to include development level namespaces in production.

The merge keeps the final array unique, to prevent accidental duplication of a namespace key. Duplication might cause other problems, but this doesn't have to be one of them. 😉 

So, all of that said, i can alter it any way you wish. i can always include or always exclude the dev namespaces, which removes the environment check. i could also flip it so the dev set is added first, then production overwrites any accidental duplicate entry... Whatever you'd prefer.

```php
/*
 * Given this structure:
 * {
 *   "autoload": {
 *     "psr-4": {
 *       "Foo\\": "foo",
 *       "Bar\\": "bar"
 *     },
 *  },
 * "autoload-dev": {
 *   "psr-4": {
 *     "Tests\\": "tests",
 *     "Bar\\": "bar-tests" (hypothetical duplicate namespace!)
 *   }
 * }
 */


// Always include dev (prod overwrites dev)
$autoloads = array_merge($composer['autoload-dev']['psr-4'], $composer['autoload']['psr-4'])
/*
 * Result:
 * {
 *   "Tests\\": "tests",
 *   "Foo\\": "foo",
 *   "Bar\\": "bar"
 * }
 */

// Always include dev (dev overwrites prod)
$autoloads = array_merge($composer['autoload']['psr-4'], $composer['autoload-dev']['psr-4'])
/*
 * Result:
 * {
 *   "Foo\\": "foo",
 *   "Tests\\": "tests",
 *   "Bar\\": "bar-tests"
 * }
 */

// Always exclude dev
$autoloads = $composer['autoload']['psr-4'];
/*
 * Result:
 * {
 *   "Foo\\": "foo",
 *   "Bar\\": "bar"
 * }
 */

// Conditionally include dev (dev overwrites prod - current submission)
$autoloads = [ $composer['autoload']['psr-4'] ];
// if not-prod add dev section and merge the two sections to a single set of namespaces
/*
 * Result (prod):
 * {
 *   "Foo\\": "foo",
 *   "Bar\\": "bar"
 * }
 * Result (not prod):
 * {
 *   "Foo\\": "foo",
 *   "Tests\\": "tests",
 *   "Bar\\": "bar-tests"
 * }
 */

// Conditionally include dev (prod overwrites dev)
if ( /* not-production */ ) {
    $autoloads = [ $composer['autoload-dev']['psr-4'] ];
}
$autoloads[] = $composer['autoload']['psr-4'];
$autoloads = array_merge(...$autoloads);
/*
 * Result (prod):
 * {
 *   "Foo\\": "foo",
 *   "Bar\\": "bar"
 * }
 * Result (not prod):
 * {
 *   "Tests\\": "tests",
 *   "Foo\\": "foo",
 *   "Bar\\": "bar"
 * }
 */
```

Perhaps i'm really over-thinking it, but it seems like an important detail, so i want to be thorough...

Reply:
I think the best approach is to never look at autoload-dev for discovery. The only time we need it is for testing of the framework itself, which we can easily fix within `Tempest\Tests\TestCase::setUp()`

Don't worry about tests breaking if you omit the autoload-dev part, I'm happy to fix that once the rest is in place :) I need to tinker around with it a bit myself, so I can't give you a ready-solution, but don't worry about it :) 

Are you ok with omitting all autoload-dev detection?

Reply:
i'm _completely_ OK with leaving the dev info out of it. 😄